cmd/config/docs/api-conventions/functions-spec.md

```diff
@@ -0,0 +1,160 @@
+apiVersion: v1
+kind: ConfigMapList
+items:
+  - apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: config1
+    data:
+      p1: v1
+      p2: v2
+  - apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: config2
+```
+
+An example using `v1/List` as input:
+
+```yaml
+apiVersion: v1
+kind: List
+items:
+  spec:
+   - apiVersion: foo-corp.com/v1
+    kind: FulfillmentCenter
+    metadata:
+        name: staging
+        metadata:
+        configFn:
+            container:
+            image: gcr.io/example/foo:v1.0.0
+    address: "100 Main St."
+  - apiVersion: rbac.authorization.k8s.io/v1
+    kind: ClusterRole
+    metadata:
+      name: namespace-reader
+    rules:
+      - resources:
+          - namespaces
+        apiGroups:
+          - ""
+        verbs:
+          - get
+          - watch
+          - list
+```
+
+In addition, a function MUST accept as input a List of kind `ResourceList` where the `configFn` field, if present, will contain the config triggering the function. Functions MAY consider this field optional so that they can be triggered in an ad-hoc fashion. In this case, the function should only look at the `items` field.
+
+An example using `config.kubernetes.io/v1beta1/ResourceList` as input:
+
+```yaml
+apiVersion: config.kubernetes.io/v1beta1
+kind: ResourceList
+functionConfig:
+  apiVersion: foo-corp.com/v1
+  kind: FulfillmentCenter
+  metadata:
+    name: staging
+    metadata:
+      configFn:
+        container:
+          image: gcr.io/example/foo:v1.0.0
+  spec:
+    address: "100 Main St."
+items:
+  - apiVersion: rbac.authorization.k8s.io/v1
+    kind: ClusterRole
+    metadata:
+      name: namespace-reader
+    rules:
+      - resources:
+          - namespaces
+        apiGroups:
+          - ""
+        verbs:
+          - get
+          - watch
+          - list
+```
+
+### Output Type
+
+A function’s output MUST be the same as the input specification above -- i.e. `ResourceList` or `List`. This is necessary to enable chaining two or more functions together in a pipeline. The serialization format of the output SHOULD match that of its input on each invocation -- e.g. if the input was a `ResourceList`, the output should also be a `ResourceList`.
+
+### Serialization Format
+
+A function MUST support YAML as a serialization format for the input and output. A function MUST use utf8 encoding (as YAML is a superset of JSON, JSON will also be supported by any conforming function).
+
+### Operations
+
+A function MAY Create, Update, or Delete any number of items in the `items` field and output the resultant list.
+
+A function MAY modify annotations with prefix `config.kubernetes.io`, but must be careful about doing so since they’re used for orchestration purposes and may interfere with other functions in the pipeline.
+
+A function SHOULD preserve comments when input serialization format is YAML. This allows for human authoring of configuration to coexist with changes made of functions.
```

Comment:
I suspect a large number of functions will not do this if written using libraries that don't preserve them.  Not sure if there is anything else to say on this though.

Reply:
SGTM